### PR TITLE
feat(editor): implement enter key strategy pattern and fix newline formatting bugs

### PR DESCRIPTION
**Pull request by mkeeda**
## Overview
This PR refactors the newline (Enter key) handling mechanism in the `RichTextEditor` by introducing the Strategy pattern (`EnterKeyStrategy`). It successfully decouples paragraph attribute inheritance logic from `RichTextState`, enabling a robust, extensible approach for handling block-level formatting like Headings and Lists upon newline insertion.

In addition to the architectural improvements, this PR addresses a number of critical bugs related to span shifts, paragraph snapping, attribute inheritance, and visual spacing that surfaced during the implementation of the new enter key strategies.

### Architecture
- **EnterKeyStrategy Pattern**: Created pure function strategies (`InheritParagraphStrategy`, `HeadingEnterStrategy`, `ListEnterStrategy`) to determine `EnterKeyResult`s (Inherit, Clear, Outdent) based on the current context without mutating state directly. Modified `ListEnterStrategy` to correctly Outdent empty Level 1 list items instead of blindly clearing them.
- **EnterKeyHandler**: Extracted span shifting and buffer mutation logic into a dedicated handler object, drastically reducing the complexity previously housed within `RichTextState.updateRichString`.

### Bug Fixes
- **Paragraph Snapping & Expansion Fixes**: Corrected boundary conditions where paragraph attributes erroneously expanded across newlines during insertions. Fixed paragraph snapping logic to safely include trailing newlines at `text.length`.
- **Newline Inheritance Bug (Outdent)**: Separated the inheritance flow for span attributes vs paragraph attributes during Outdent operations, fixing a bug where styles were incorrectly re-applied to the previous line.
- **Backspace Merge Fix**: Adjusted `shiftSpans` to explicitly detect when two paragraphs merge due to a deleted newline, dropping the stale `ParagraphAttribute`s of the bottom paragraph so the top paragraph correctly retains its original styling (fixes nested list indentation loss on backspace).
- **Outdent Visual Gap Fix**: Updated `ComposeParagraphWorkarounds` to conditionally replace trailing newlines with zero-width non-breaking spaces (`\uFEFF`) when they sit on a paragraph style boundary, effectively preventing Compose from double-rendering trailing empty lines after an Outdent.

### Testing
- **Integration Tests**: Migrated `RichTextStateTest` logic to `RichTextEditorTest` to ensure that boundary condition tests execute real text insertions (e.g. sequentially entering `\n` then text) rather than simulating paste actions, closely mimicking real user keyboard input.
- **Visual Regression**: Updated Roborazzi screenshot tests (`RichTextEditorScreenshotTest`) to accurately reflect the corrected list outdent behavior.

---

## Future Architecture Proposal
Many of the bugs fixed in this PR (especially those related to span boundary index math and paragraph merging) stem from the inherent complexity of mutating our current **Flat Model** (where styles are mapped to string indices via `IntRange`).

To completely eliminate these classes of bugs and prepare for advanced features like inline media or tables, we have published a proposal to migrate to an **AST / Block-based Document Model**.

**See Issue #53 for details:** [Architectural Proposal: Migrate from Flat Model to AST/Block-based Document Model](https://github.com/mkeeda/arranger/issues/53)

### CODE REVIEW

File: richtext-editor/src/main/kotlin/dev/mkeeda/arranger/richtext/editor/EnterKeyHandler.kt [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        val blockKeysToClear = context.currentAttributes.keys.filterIsInstance<BlockTypeAttributeKey<*>>()
+        val newParagraphStart = insertedRange.last + 1
+        val snappedRange = (newParagraphStart..newParagraphStart).snapToParagraphs(buffer.toString())
+        val tempBuffer = dev.mkeeda.arranger.richtext.editor.RichTextBuffer(spans, buffer)
+        blockKeysToClear.forEach { key ->
+            @Suppress("UNCHECKED_CAST")
+            tempBuffer.removeParagraphAttribute(key as ParagraphAttributeKey<Any>, snappedRange)
+        }
+        val clearedSpans = tempBuffer.spans
+
```
Comment (mkeeda) [inline richtext-editor/src/main/kotlin/dev/mkeeda/arranger/richtext/editor/EnterKeyHandler.kt:83]: The fully-qualified class name `dev.mkeeda.arranger.richtext.editor.RichTextBuffer` is used here instead of a proper import. This violates the project's import style guide which says to avoid fully-qualified names and prefer short import statements.

Please add an import for `RichTextBuffer` and use it directly:

```kotlin
// Add at top of file:
import dev.mkeeda.arranger.richtext.editor.RichTextBuffer

// Then use:
val tempBuffer = RichTextBuffer(spans, buffer)
```

File: richtext-editor/src/main/kotlin/dev/mkeeda/arranger/richtext/editor/EnterKeyHandler.kt [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        // 3. Apply the outdented attributes to the current paragraph
+        val snappedRange = context.paragraphRange.snapToParagraphs(buffer.toString())
+        val tempBuffer = dev.mkeeda.arranger.richtext.editor.RichTextBuffer(revertedSpans, buffer)
+
+        // Clear all previous block attributes to ensure we don't leak removed attributes (e.g., when completely outdenting to empty)
+        val blockKeysToClear = context.currentAttributes.keys.filterIsInstance<BlockTypeAttributeKey<*>>()
+        blockKeysToClear.forEach { key ->
+            @Suppress("UNCHECKED_CAST")
+            tempBuffer.removeParagraphAttribute(key as ParagraphAttributeKey<Any>, snappedRange)
```
Comment (mkeeda) [inline richtext-editor/src/main/kotlin/dev/mkeeda/arranger/richtext/editor/EnterKeyHandler.kt:122]: Same fully-qualified name issue as above — `dev.mkeeda.arranger.richtext.editor.RichTextBuffer` appears again on line 111. Once the import is added, both occurrences can be simplified.

File: richtext-editor/src/main/kotlin/dev/mkeeda/arranger/richtext/editor/RichTextState.kt [hunk 3]
```diff
@@ -319,21 +326,107 @@ public class RichTextState(initialText: RichString) {
         this.spans = newSpans.resnapParagraphSpans(buffer.toString())
         clearTypingAttributes()
     }
+
+    private fun handleNewlineInsertion(
+        insertedText: String,
```
Comment (mkeeda) [inline richtext-editor/src/main/kotlin/dev/mkeeda/arranger/richtext/editor/RichTextState.kt:305]: Fully-qualified names `dev.mkeeda.arranger.richtext.EnterKeyContext` and `dev.mkeeda.arranger.richtext.InheritParagraphStrategy` are used here. Per the project style guide, these should be shortened via import statements.

```kotlin
// Add imports:
import dev.mkeeda.arranger.richtext.EnterKeyContext
import dev.mkeeda.arranger.richtext.InheritParagraphStrategy

// Then use:
val context = EnterKeyContext(
    text = buffer.toString(),
    cursorPosition = range.min,
    paragraphRange = paragraphRange,
    currentAttributes = attrsBeforeCursor,
)

val strategy =
    paragraphAttrKeys
        .map { it.enterKeyStrategy }
        .firstOrNull { it != InheritParagraphStrategy }
        ?: InheritParagraphStrategy
```

File: richtext-editor/src/main/kotlin/dev/mkeeda/arranger/richtext/editor/EnterKeyHandler.kt [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        spans: List<RichSpan>,
+        buffer: TextFieldBuffer,
+        insertedRange: IntRange,
+    ): List<RichSpan> {
+        // 1. Remove the inserted newline from the buffer
+        buffer.replace(insertedRange.first, insertedRange.last + 1, "")
+
+        // 2. Shift spans back to account for the removed newline
+        val revertedSpans =
+            spans.shiftSpans(
```
Comment (mkeeda) [inline richtext-editor/src/main/kotlin/dev/mkeeda/arranger/richtext/editor/EnterKeyHandler.kt:107]: `applyOutdent` mutates `buffer` directly via `buffer.replace(...)`. This is a significant side effect that is invisible from the call site in `apply()`. A reader of `EnterKeyHandler.apply()` would not expect that certain result types cause the buffer to be mutated, while others do not.

Please add a KDoc comment to `applyOutdent` (and ideally to `apply()` as well) explicitly documenting that this method mutates the `buffer`:

```kotlin
/**
 * Applies the Outdent result by removing the inserted newline from [buffer] and shifting spans accordingly.
 *
 * **Note:** This method mutates [buffer] directly by removing the character at [insertedRange].
 */
private fun applyOutdent(...)
```

File: richtext-editor/src/main/kotlin/dev/mkeeda/arranger/richtext/editor/RichTextState.kt [hunk 5]
```diff
@@ -283,34 +308,16 @@ public class RichTextState(initialText: RichString) {
...
-                            updatedSpans.mergeSpan(
-                                RichSpan(
-                                    range = snappedRange,
-                                    attributes = paragraphAttr,
-                                ),
-                            )
-                    }
+                    updatedSpans =
+                        handleNewlineInsertion(
+                            insertedText = insertedText,
```
Comment (mkeeda) [inline richtext-editor/src/main/kotlin/dev/mkeeda/arranger/richtext/editor/RichTextState.kt:287]: The `handleNewlineInsertion()` function builds `attrsBeforeCursor` using the same span-fold pattern that already exists in `computeCurrentAttributes()`. This logic is now duplicated in two places, which increases maintenance risk if the inheritance rules change in the future.

Consider extracting a shared helper (e.g. `collectAttributesAt(spanIndex: Int, paragraphIndex: Int)`) that both `computeCurrentAttributes()` and `handleNewlineInsertion()` delegate to.

File: richtext-editor/src/test/kotlin/dev/mkeeda/arranger/richtext/editor/RichTextEditorTest.kt [hunk 6]
```diff
@@ -254,4 +259,181 @@ class RichTextEditorTest {
...
+        // The span for the empty line at index 14..14 should be Level 1.
+        val textAfterEnters = "Item 1\nItem 2\n"
+
+        // Select the last newline character
+        composeTestRule.onNodeWithText(
+            textAfterEnters,
+        ).performTextInputSelection(TextRange(textAfterEnters.length - 1, textAfterEnters.length))
+
+        // Replace it with empty string (simulating Backspace)
+        composeTestRule.onNodeWithText(textAfterEnters).performTextInput("")
```
Comment (mkeeda) [inline richtext-editor/src/test/kotlin/dev/mkeeda/arranger/richtext/editor/RichTextEditorTest.kt:429]: The current backspace simulation — selecting the character and replacing it with an empty string via `performTextInput("")` — may not accurately reproduce what happens when a user presses the physical Backspace key. Specifically, IME-based single-character deletion triggers a different code path in `TextFieldBuffer` than a selection-replace, so the test may not be exercising the same `shiftSpans` logic as a real delete event.

Consider using `performKeyInput` with `Key.Backspace` to more faithfully simulate a real key event:

```kotlin
import androidx.compose.ui.test.performKeyInput
import androidx.compose.ui.input.key.Key
import androidx.compose.ui.input.key.pressKey

// Move cursor to end, then press Backspace
composeTestRule.onNodeWithText(textAfterEnters)
    .performTextInputSelection(TextRange(textAfterEnters.length))
composeTestRule.onNodeWithText(textAfterEnters)
    .performKeyInput { pressKey(Key.Backspace) }
```

If `performKeyInput` is not available or not reliable in this Robolectric environment, please add a comment explaining the trade-off so future contributors understand why `performTextInput("")` was chosen over a more direct key simulation.

**Review by mkeeda — COMMENTED** [review]
(AI review)

Great work on this PR! The Strategy pattern refactoring is well-structured and makes the Enter key behavior extensible. The `EnterKeyContext` + `EnterKeyResult` sealed interface design is clean and the KDoc on the public API is thorough. CI passes.

A few items to address:

**Must fix:**
- Fully-qualified class names (`dev.mkeeda.arranger.richtext.editor.RichTextBuffer`, `dev.mkeeda.arranger.richtext.EnterKeyContext`, `dev.mkeeda.arranger.richtext.InheritParagraphStrategy`) are used in `EnterKeyHandler.kt` and `RichTextState.kt`. These should be shortened via import statements per the project style guide.
- `applyOutdent` mutates `buffer` directly as a side effect, which is invisible from the call site. This should be documented in KDoc.

**Suggestions:**
- The span-fold attribute collection logic in `handleNewlineInsertion` is duplicated from `computeCurrentAttributes`. Consider extracting a shared helper to reduce maintenance risk.
- The Backspace simulation in the test (`performTextInput("")` after selecting a character) may not reflect the actual IME deletion path. Consider `performKeyInput { pressKey(Key.Backspace) }` or add a comment explaining the trade-off.

**Comment by mkeeda** [comment]
I have addressed all the feedback points raised in this code review:
- Added missing imports for `RichTextBuffer`, `EnterKeyContext`, and `InheritParagraphStrategy` to adhere to the import style guide.
- Added KDocs documenting the side-effects of `buffer` mutation in `EnterKeyHandler.apply()` and `applyOutdent()`.
- Extracted the shared span-attribute folding logic into a new helper function `collectInheritedAttributes()` to DRY up `RichTextState.kt`.
- Attempted to use `performKeyInput { pressKey(Key.Backspace) }` in `RichTextEditorTest.kt`. However, due to Robolectric's experimental limitations with `TextFieldBuffer` key events, I reverted to `performTextInput("")` and added an explanatory comment as suggested.

The changes have been tested, formatted with Spotless, and pushed.
